src/client/metric.rs

```diff
@@ -286,11 +291,170 @@ impl<T: MetricType + Clone> Metric<T> {
         self.mmap_view = mmap_view;
     }
 
-    pub (super) fn write_val(&self, cursor: &mut Cursor<&mut [u8]>) -> io::Result<()> {
-        self.val.write_to_writer(cursor)
+    pub (super) fn write_val<W: Write>(&self, writer: &mut W) -> io::Result<()> {
+        self.val.write_to_writer(writer)
     }
 }
 
+pub struct InstanceMetric<T> {
+    pub (super) metrics: HashMap<String, Metric<T>>,
```

Comment:
Although, on a second look the amount of duplicated data seems small enough, so that might be overkill